osu.Game/Screens/Spectate/SpectatorScreen.cs

```diff
@@ -99,51 +99,55 @@ private void beatmapUpdated(BeatmapSetInfo beatmapSet)
         {
             foreach ((int userId, _) in userMap)
             {
-                if (!playingUserStates.TryGetValue(userId, out var userState))
+                if (!userStates.TryGetValue(userId, out var userState))
                     continue;
 
                 if (beatmapSet.Beatmaps.Any(b => b.OnlineID == userState.BeatmapID))
                     updateGameplayState(userId);
             }
         }
 
-        private void onPlayingUserStatesChanged(object sender, NotifyDictionaryChangedEventArgs<int, SpectatorState> e)
+        private void onUserStatesChanged(object sender, NotifyDictionaryChangedEventArgs<int, SpectatorState> e)
         {
             switch (e.Action)
             {
                 case NotifyDictionaryChangedAction.Add:
+                case NotifyDictionaryChangedAction.Replace:
                     foreach ((int userId, var state) in e.NewItems.AsNonNull())
-                        onUserStateAdded(userId, state);
+                        onUserStateChanged(userId, state);
                     break;
 
                 case NotifyDictionaryChangedAction.Remove:
-                    foreach ((int userId, var _) in e.OldItems.AsNonNull())
-                        onUserStateRemoved(userId);
-                    break;
-
-                case NotifyDictionaryChangedAction.Replace:
-                    foreach ((int userId, var _) in e.OldItems.AsNonNull())
-                        onUserStateRemoved(userId);
-
-                    foreach ((int userId, var state) in e.NewItems.AsNonNull())
-                        onUserStateAdded(userId, state);
+                    foreach ((int userId, SpectatorState state) in e.OldItems.AsNonNull())
+                        onUserStateRemoved(userId, state);
                     break;
             }
         }
 
-        private void onUserStateAdded(int userId, SpectatorState state)
+        private void onUserStateChanged(int userId, SpectatorState newState)
         {
-            if (state.RulesetID == null || state.BeatmapID == null)
+            if (newState.RulesetID == null || newState.BeatmapID == null)
                 return;
 
             if (!userMap.ContainsKey(userId))
                 return;
 
-            Schedule(() => OnUserStateChanged(userId, state));
-            updateGameplayState(userId);
+            switch (newState.State)
+            {
+                case SpectatingUserState.Passed:
+                    // Make sure that gameplay completes to the end.
+                    if (gameplayStates.TryGetValue(userId, out var gameplayState))
+                        gameplayState.Score.Replay.HasReceivedAllFrames = true;
+                    break;
+
+                case SpectatingUserState.Playing:
+                    Schedule(() => OnUserStateChanged(userId, newState));
+                    updateGameplayState(userId);
```

Comment:
Was confused to see a method named `updateGameplayState` called only when `newState.State == Playing`, but seems that method is rather meant for starting gameplay?

Perhaps it should be renamed to `startGameplay` to better reflect what it's doing.